src/lib/core/data-table/data-table.ts

```diff
@@ -0,0 +1,323 @@
+}
+
+/**
+ * A data table that connects with a data source to retrieve data and renders
+ * a header row and data rows. Updates the rows when new data is provided by the data source.
+ */
+@Component({
+  selector: 'cdk-table',
+  template: `
+    <ng-container cdkHeaderRowPlaceholder></ng-container>
+    <ng-container cdkRowPlaceholder></ng-container>
+  `,
+  host: {
+    'class': 'cdk-table',
+    'role': 'grid' // TODO: Allow the user to choose either grid or treegrid
+  },
+  encapsulation: ViewEncapsulation.None,
+  changeDetection: ChangeDetectionStrategy.OnPush,
+})
+export class CdkTable {
+  /**
+   * Provides a stream containing the latest data array to render. Influenced by the table's
+   * stream of view window (what rows are currently on screen).
+   */
+  @Input() dataSource: DataSource<any>;
+
+  /**
+   * Stream containing the latest information on what rows are being displayed on screen. Offered
+   * to the data source as a heuristic of what data should be provided.
+   */
+  // TODO: Remove max value as the end index and instead calculate the view on init and scroll.
+  viewChange = new BehaviorSubject<CollectionViewer>({start: 0, end: Number.MAX_VALUE});
+
+  /**
+   * Map of all the user's defined columns identified by name.
+   * Contains the header and data-cell templates.
+   */
+  private _columnMap = new Map<ConstrainDOMString,  CdkColumnDef>();
+
+  // Placeholders within the table's template where the header and data rows will be inserted.
+  @ViewChild(CdkRowPlaceholder) rowPlaceholder: CdkRowPlaceholder;
+  @ViewChild(CdkHeaderRowPlaceholder) headerRowPlaceholder: CdkHeaderRowPlaceholder;
+
+  /**
+   * The column definitions provided by the user that contain what the header and cells should
+   * render for each column.
+   */
+  @ContentChildren(CdkColumnDef) columnDefinitions: QueryList<CdkColumnDef>;
+
+  /** Template used as the header container. */
+  @ContentChild(CdkHeaderRowDef) headerDefinition: CdkHeaderRowDef;
+
+  /** Set of templates that used as the data row containers. */
+  @ContentChildren(CdkRowDef) rowDefinitions: QueryList<CdkRowDef>;
+
+  constructor(private _changeDetectorRef: ChangeDetectorRef) {}
+
+  ngOnDestroy() {
+    // TODO: Disconnect from the data source so that it can unsubscribe from its streams.
+  }
+
+  ngOnInit() {
+    // TODO: Setup a listener for scroll events and emit the calculated view to this.viewChange
+  }
+
+  ngAfterContentInit() {
+    this.columnDefinitions.forEach(columnDef => this._columnMap.set(columnDef.name, columnDef));
+  }
+
+  ngAfterViewInit() {
+    // TODO: Re-render the header when the header's columns change.
+    this.renderHeaderRow();
+
+    // TODO: Re-render rows when their list of columns change.
+    // TODO: If the data source is not present after view init, connect it when it is defined.
+    const connectFn = this.dataSource.connectTable.bind(this.dataSource);
+    this.viewChange.let(connectFn).subscribe((rowsData: any[]) => {
```

Comment:
I think its a legacy from prior designs, I don't think we need let at all anymore and can do what you're saying.